proxy/envoy/ingress.go

```diff
@@ -0,0 +1,244 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package envoy
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/golang/glog"
+
+	"istio.io/manager/model"
+
+	"k8s.io/client-go/pkg/apis/extensions/v1beta1"
+)
+
+type ingressWatcher struct {
+	agent   Agent
+	mesh    *MeshConfig
+	ingress model.IngressRegistry
+}
+
+// NewWatcher creates a new ingress watcher instance with an agent
+func NewIngressWatcher(ingress model.IngressRegistry, ctl model.Controller, mesh *MeshConfig) (Watcher, error) {
+	out := &ingressWatcher{
+		agent:   NewAgent(mesh.BinaryPath, mesh.ConfigPath),
+		mesh:    mesh,
+		ingress: ingress,
+	}
+
+	if err := ctl.AppendIngressHandler(out.notify); err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+func (w *ingressWatcher) notify(ingress *v1beta1.Ingress, event model.Event) {
+	config, err := w.generateConfig()
+	if err != nil {
+		glog.Warningf("Failed to generate Envoy configuration: %v", err)
+		return
+	}
+
+	current := w.agent.ActiveConfig()
+	if reflect.DeepEqual(config, current) {
+		glog.V(2).Info("Configuration is identical, skipping reload")
+		return
+	}
+
+	// TODO: add retry logic
+	if err := w.agent.Reload(config); err != nil {
+		glog.Warningf("Envoy reload error: %v", err)
+		return
+	}
+
+	// Add a short delay to de-risk a potential race condition in envoy hot reload code.
+	// The condition occurs when the active Envoy instance terminates in the middle of
+	// the Reload() function.
+	time.Sleep(256 * time.Millisecond)
+}
+
+func (w *ingressWatcher) generateConfig() (*Config, error) {
+	// Map a domain name to a VirtualHost
+	vhostsMap := make(map[string]VirtualHost)
+
+	// Map a cluster name to a Cluster
+	clustersMap := make(map[string]Cluster)
+
+	// Function to add a rule for a particular host, path and backend.
+	// This will later be invoked for each ingress rule, as well as ingress default backend
+	addRule := func(host, path, namespace string, backend *v1beta1.IngressBackend) {
+
+		// If host is unspecified, use Envoy's wildcard domain "*"
+		domain := host
+		if domain == "" {
+			domain = "*"
+		}
+
+		// Fetch domain's VirtualHost, or create one if doesn't exist yet
+		vhost, ok := vhostsMap[domain]
+		if !ok {
+			vhost := VirtualHost{
+				Name:    domain,
+				Domains: []string{domain},
+			}
+			vhostsMap[domain] = vhost
+		}
+
+		// Fetch target cluster, or create one if doesn't exist yet
+		clusterName := clusterName(namespace, backend)
+		cluster, ok := clustersMap[clusterName]
+		if !ok {
+			clustersMap[clusterName] = Cluster{
+				Name:             clusterName,
+				ServiceName:      clusterName,
+				Type:             "strict_dns", // TODO: Change to SDS
+				LbType:           DefaultLbType,
+				ConnectTimeoutMs: DefaultTimeoutMs,
+				Hosts:            clusterHosts(namespace, backend),
```

Comment:
don't we have to set the cluster `"features": "http2"`, since we would have Envoy backends? If this is for istio only, then by default (and for first cut) I think we should set to features to "http2", since all http communication happens over Envoy to Envoy. If not specified, Envoy falls back to http/1.1, nullifying a lot of benefits that we get out of http/2.